### PR TITLE
Add implementation details in DG

### LINKED ISSUE #112: Colin: Update DG
TODO: when you've decided which feature you are going to document, change the subject of the title.
e.g. `NAME: Update DG` -> `FEATURE: Update DG`

Expected length: 1+ page per person
The description can contain things such as,
* How the feature is implemented (or is going to be implemented).
* Why it is implemented that way.
* Alternatives considered.

See [reference](https://nus-cs2103-ay2021s2.github.io/website/schedule/week10/project.html) for more details.

### PR DESCRIPTION
**Pull request by colintkn**
Closes #112 
Implementation details for ResidentRoom.

### CODE REVIEW

**Review by benedictkhoomw — CHANGES_REQUESTED** [review]
Your PR as-is appears to be committing the contents of a `.gradle` folder but I don't think these are supposed to be committed. For example, our `.gitignore` lists `/.gradle/`. Perhaps `.gradle/` was accidentally moved into the `docs/` folder.

File: docs/DeveloperGuide.md [hunk 1]
```diff
@@ -132,6 +132,41 @@ Classes used by multiple components are in the `seedu.addressbook.commons` packa
...
+  Only if the `r/` field is specified, then `UniqueRoomList` will be checked to
+  see if the room exists. If the room does not exist, return an error, 
+  else allocate the resident to room in `ResidentRoomList`.  
+* `redit` is the main command to allocate a resident to a room. e.g. `redit 3 r/03-100`.
+  ![Activity Diagram of Allocation](images/ResidentRoomAllocationDiagram.png)
+* `rlist` and `rfind` will lookup `ResidentRoomList` for the room assigned. If `ResidentRoomList` exists, it will 
+return the allocated room, else return `Room unallocated`. 
+  
+The following implementation alternatives were considered: 
+* Resident-Room allocation can be modelled as a Parent-Child where the parent is the `Room` and the child is the `Resident`.
```
Comment (stein414) [inline docs/DeveloperGuide.md:168]: Perhaps the details not choosing the alternative could be added to be clearer why it was not used in the end

File: docs/DeveloperGuide.md [hunk 2]
```diff
@@ -132,6 +132,41 @@ Classes used by multiple components are in the `seedu.addressbook.commons` packa
 ## **Implementation**
 
 This section describes some noteworthy details on how certain features are implemented.
+### Resident-Room allocation feature 
+The allocation feature is facilitated by `ResidentRoom` which is a pair value of 
+valid`[Resident, Room]` that represents an existing resident allocated to a room.
```
Comment (stein414) [inline docs/DeveloperGuide.md:137]: This is a small one but maybe add a space between valid and '`'

File: docs/DeveloperGuide.md [hunk 3]
```diff
@@ -132,6 +132,41 @@ Classes used by multiple components are in the `seedu.addressbook.commons` packa
...
+A `ResidentRoomList` is a supplementary class that tracks all the `ResidentRoom` assignments. It should support
+the following operations. 
+* `ResidentRoomList#allocate()` - Adds a `ResidentRoom` new allocation 
+* `ResidentRoomList#deallocate()` - Removes an existing `ResidentRoom` allocation.
+* `ResidentRoomList#isRoomOccupied()` - Checks if a room is occupied. 
+* `ResidentRoomList#isResidentOccupied()` - Checks if resident is allocated a room.
+
+The following are the dependencies between a `Room` and `ResidentRoomList`: 
+* `oadd` will create a room that is not occupied by default. 
+* `oedit` cannot edit the occupancy status directly. Occupancy status can only be done by allocation and deallocation
```
Comment (stein414) [inline docs/DeveloperGuide.md:151]: Maybe reword the second sentence as it feels a little weird to read.

**Review by stein414 — CHANGES_REQUESTED** [review]
It seems good just need to consider some changes

File: docs/DeveloperGuide.md [hunk 4]
```diff
@@ -132,6 +132,53 @@ Classes used by multiple components are in the `seedu.addressbook.commons` packa
...
+* Checks that no other room is allocated to `John Tan`.  
+* Checks that there is room `03-100` is occupied by anyone. 
+
+    **If all the above is true,**
+* Set the `ROOM` of `John Tan` to be `03-100`.
+* Set the `OCCUPATION_STATUS` of a room to `Y`.
+
+  ![Activity Diagram of Allocation](images/ResidentRoomAllocationDiagram.png)
+
+The `dealloc` command will do the following:
```
Comment (DrWala) [inline docs/DeveloperGuide.md:168]: Perhaps it would be helpful here to link to the cascading impact on commands like `rdel`, `redit`, `odel` and `oedit`, which will need to be modified to block editing until deallocation happens. It would help explain the design decisions made on those commands.

File: docs/UserGuide.md [hunk 5]
```diff
@@ -110,10 +109,11 @@ Examples:
 
 Edits the existing resident record at a specified index.
 
-Format: `redit INDEX [n/NAME] [p/PHONE_NUMBER] [e/EMAIL] [r/ROOM]`
+Format: `redit INDEX [n/NAME] [p/PHONE_NUMBER] [e/EMAIL]`
 * `INDEX` refers to the index number shown in the displayed resident list. `INDEX` **must be a positive integer: 1, 2, 3, …**.
 * At least one of the optional fields must be provided.
 * Existing values will be updated to the input values.
+* `redit` will be blocked if the resident has an active room allocation. `dealloc` before making further edits.
```
Comment (DrWala) [inline docs/UserGuide.md:116]: Just a syntactic/clarity thing: Perhaps update the sentence to say "Run `dealloc` on the resident you are editing first." and then provide an example command

**Review by DrWala — CHANGES_REQUESTED** [review]

File: docs/diagrams/ResidentRoomAllocationDiagram.puml [hunk 6]
```diff
@@ -0,0 +1,60 @@
+@startuml
+'https://plantuml.com/activity-diagram-beta
+
+start
+if (UniqueRoomList.contains()) then (true)
```
Comment (benedictkhoomw) [inline docs/diagrams/ResidentRoomAllocationDiagram.puml:5]: This activity diagram appears to have the condition within the branching diamond. However, this may not meet requirements for this module. Consider moving the branching conditions out of the diamond as guard conditions in square brackets.

File: docs/DeveloperGuide.md (hunk 1 above)
Comment (benedictkhoomw) [reply docs/DeveloperGuide.md:180]: Agreed. It may also be good to list out pros and cons for each possible implementation.

File: docs/diagrams/ResidentRoomAllocationDiagram.puml [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+            stop
+    endif
+    else (false)
+        stop
+   endif
+
+else (false)
+    stop
+
+/'
```
Comment (benedictkhoomw) [inline docs/diagrams/ResidentRoomAllocationDiagram.puml:29]: This looks like commented-out code. Was this intended to be here?

**Review by benedictkhoomw — CHANGES_REQUESTED** [review]
Looks mostly good. Just two issues with the activity diagram. One is syntactical and probably should be fixed before merging; it is mentioned in a comment on the `.puml` file. The other is semantic and is detailed below.

The activity diagram seems a bit clunky in that it is modelling a complex boolean expression. However, there may not be a need to have five layers since it seems to be one decision. That is,

If resident exists and room exists and ..., then do the allocation. Otherwise, show an error message. 

Perhaps the five branches could be collapsed into one branch with the necessarily complex guard condition on one flow and `[else]` on the other flow. The former flow could lead to the allocation actions while the latter flow could lead to the end node. This approach could simplify the diagram and make your intention clearer.

EDIT: consider adding a sequence diagram of either the `alloc` or `dealloc` command. This could better clarify how this allocation system interacts with other components of SunRez. Additionally, it gives you another diagram to show off your documentation skills with :)

File: docs/diagrams/ResidentRoomAllocationDiagram.puml (hunk 7 above)
Comment (colintkn) [reply docs/diagrams/ResidentRoomAllocationDiagram.puml:29]: removed this chunk of code. Thank for pointing out.

File: docs/diagrams/ResidentRoomAllocationDiagram.puml (hunk 6 above)
Comment (colintkn) [reply docs/diagrams/ResidentRoomAllocationDiagram.puml:5]: Noted. Added actions to represent the checks. And used guard clauses.

File: docs/UserGuide.md (hunk 5 above)
Comment (colintkn) [reply docs/UserGuide.md:116]: Ok. Added the example command to the userGuide.

File: docs/DeveloperGuide.md (hunk 4 above)
Comment (colintkn) [reply docs/DeveloperGuide.md:168]: Added this point. Thank you.

File: docs/DeveloperGuide.md [hunk 8]
```diff
@@ -132,6 +132,60 @@ Classes used by multiple components are in the `seedu.addressbook.commons` packa
...
+Example: `dealloc n/John Tan r/03-100`
+* Checks that `John Tan` exists.
+* Checks that `03-100` exists.
+* Checks that `John Tan` has been allocated to `03-100`.
+
+    **If all the above is true,**
+* Set the `ROOM` of `John Tan` to be `Room unallocated`.
+* Set the `OCCUPATION_STATUS` of a room to `N`.
+
+The following implementation alternatives were considered:
```
Comment (DrWala) [inline docs/DeveloperGuide.md:179]: I think the formatting of this should reference AB3: https://se-education.org/addressbook-level3/DeveloperGuide.html#design-consideration

File: docs/diagrams/ResidentRoomAllocationDiagram.puml [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+    if () then ([Resident exists])
+        : Check if Resident or Room is already allocated;
+        if () then ([None allocated])
+            : Set the room number in Resident;
+            : Set the occupancy as 'Y' in Room;
+            stop
+        else ([Resident or/and Room is allocated])
+            stop
+    endif
+    else ([Resident does not exist])
```
Comment (DrWala) [inline docs/diagrams/ResidentRoomAllocationDiagram.puml:17]: For consistency with other activity diagrams, perhaps state the actual condition in the "happy flow" and "else" for the flow that terminates.

**Review by DrWala — CHANGES_REQUESTED** [review]
Some minor nits to fix

**Review by benedictkhoomw — CHANGES_REQUESTED** [review]
Looks good except for a few nits.

EDIT: for consistency, could you move your `.puml` and `.png` files into their own folder? e.g. `[path]/diagrams/allocation/blah.puml`

**Review by benedictkhoomw — APPROVED** [review]
LGTM

**Review by DrWala — APPROVED** [review]
LGTM.

**Review by stein414 — APPROVED** [review]
LGTM
